Zend/zend_builtin_functions.c

```diff
@@ -637,26 +637,21 @@ ZEND_FUNCTION(get_called_class)
    Retrieves the parent class name for object or class or current scope or false if not in a scope. */
 ZEND_FUNCTION(get_parent_class)
 {
-	zval *arg;
+	zend_string *str = NULL;
+	zend_object *object = NULL;
 	zend_class_entry *ce = NULL;
 
-	if (zend_parse_parameters(ZEND_NUM_ARGS(), "|z", &arg) == FAILURE) {
-		RETURN_THROWS();
-	}
+	ZEND_PARSE_PARAMETERS_START(0, 1)
+		Z_PARAM_OPTIONAL
+		Z_PARAM_STR_OR_OBJ(str, object)
+	ZEND_PARSE_PARAMETERS_END();
 
-	if (!ZEND_NUM_ARGS()) {
+	if (object) {
+		ce = object->ce;
+	} else if (str) {
+		ce = zend_lookup_class(str);
```

Comment:
I'm a bit concerned about this. This means we now call zend_lookup_class(), which also includes an autoloader call, on things like numbers. This seems potentially problematic. Especially if we consider how this is going to extend other similar functions, like method_exists(). The latter could historically (prior to PHP 8) be called on non-strings/objects and just returned false in that case. And people did do that (based on Symfony), and these calls would now start invoking the autoloader.

As such, I think we really should be treating these "object or class name" arguments differently. Either by just handling the types strictly, as method_exists/property_exists now do, or by actually adding a special fastzpp case for it (we already have one for just class names). In that case we'd start throwing on unknown classes though, rather than just returning false.

Reply:
@nikic I implemented a `class name or object` ZPP macro. After pushing the change, I've found out that it currently has some issues, but before I start fixing them (e.g. it returns the class name rather than the `ce`), I'd like to be sure if it's the right way to go. Do you think this macro is a good idea? Or should we just use your other suggestion?

Reply:
@nikic I've started to look into the resource migration of openssl. As it turned out, a `string|object-of-type` ZPP macro will be needed there (e.g. most X.509 related functions would need it), so my question is if I could also add this macro besides the `class name or object` in this PR?

Reply:
Seems like a reasonable addition for that use-case. Let's make a separate PR though.